Mathlib/Analysis/Normed/Group/SemiNormedGroup.lean

```diff
@@ -0,0 +1,297 @@
+
+/--porting note: originally empty, which didn't work. Notation for composition changed?-/
+instance : Limits.HasZeroMorphisms.{u, u + 1} SemiNormedGroup where
+  Zero _ _ := NormedAddGroupHom.zero
+  comp_zero _ _ := rfl
+  zero_comp _ _ _ f := NormedAddGroupHom.comp_zero f
+
+
+@[simp]
+theorem zero_apply {V W : SemiNormedGroup} (x : V) : (0 : V ⟶ W) x = 0 :=
+  rfl
+#align SemiNormedGroup.zero_apply SemiNormedGroup.zero_apply
+
+/--Porting note: Had to add some proof steps -/
+theorem isZero_of_subsingleton (V : SemiNormedGroup) [Subsingleton V] : Limits.IsZero V := by
+  refine' ⟨fun X => ⟨⟨⟨0⟩, fun f => _⟩⟩, fun X => ⟨⟨⟨0⟩, fun f => _⟩⟩⟩
+  · ext x
+    change V at x
+    have  := Subsingleton.elim (x : V) (0 : V)
+    rw [this]
+    suffices f 0 = (0 : V⟶ X) 0 by convert this
+    simp [map_zero]
+  · ext
+    apply @Subsingleton.elim V _
+#align SemiNormedGroup.is_zero_of_subsingleton SemiNormedGroup.isZero_of_subsingleton
+
+instance hasZeroObject : Limits.HasZeroObject SemiNormedGroup.{u} :=
+  ⟨⟨of PUnit, isZero_of_subsingleton _⟩⟩
+#align SemiNormedGroup.has_zero_object SemiNormedGroup.hasZeroObject
+
+/--Porting Note: Added to make iso_isometry_of_normNoninc work-/
+instance toAddMonoidHomClass {V W : SemiNormedGroup} : AddMonoidHomClass (V ⟶ W) V W where
+  coe := (forget SemiNormedGroup).map
+  coe_injective' := fun f g h => by cases f; cases g; congr
+  map_add f := f.map_add'
+  map_zero f := (AddMonoidHom.mk' f.toFun f.map_add').map_zero
+
+theorem iso_isometry_of_normNoninc {V W : SemiNormedGroup} (i : V ≅ W) (h1 : i.hom.NormNoninc)
+    (h2 : i.inv.NormNoninc) : Isometry i.hom := by
+  apply AddMonoidHomClass.isometry_of_norm i.hom
+  intro v
+  apply le_antisymm (h1 v)
+  have h3 : v = i.inv (i.hom v) := by
+    change v = (forget _).map i.inv ((forget _).map i.hom v)
+    simp only [FunctorToTypes.map_inv_map_hom_apply]
+  calc
+    ‖v‖ = ‖i.inv (i.hom v)‖ := by rw [← h3]
+    _ ≤ ‖i.hom v‖ := h2 _
+
+#align SemiNormedGroup.iso_isometry_of_norm_noninc SemiNormedGroup.iso_isometry_of_normNoninc
+
+end SemiNormedGroup
+
+/-- `SemiNormedGroup₁` is a type synonym for `SemiNormedGroup`,
+which we shall equip with the category structure consisting only of the norm non-increasing maps.
+-/
+def SemiNormedGroup₁ : Type (u + 1) :=
+  Bundled SeminormedAddCommGroup
+#align SemiNormedGroup₁ SemiNormedGroup₁
+
+namespace SemiNormedGroup₁
+
+--Porting Note: hasCoeToSort doesn't exist anymore
+--instance : CoeSort SemiNormedGroup₁ (Type u) :=
+--  Bundled.hasCoeToSort
+
+instance : CoeSort SemiNormedGroup₁ (Type _) where
+  coe X := X.α
+
+instance : LargeCategory.{u} SemiNormedGroup₁ where
+  Hom X Y := { f : NormedAddGroupHom X Y // f.NormNoninc }
+  id X := ⟨NormedAddGroupHom.id X, NormedAddGroupHom.NormNoninc.id⟩
+  comp {X Y Z} f g := ⟨g.1.comp f.1, g.2.comp f.2⟩
+
+/-- Porting Note: Added-/
+instance {X Y : SemiNormedGroup₁} : CoeFun (X ⟶ Y) fun _ => X → Y where
+  coe (f : X ⟶ Y) := f.1
```

Comment:
Probably we want a `FunLike` instance on `X ⟶ Y`, rather than this coercion. Sorry, I've been meaning to think about all these concrete categories properly again, but am still uncertain about the right design. Would you mind trying out `FunLike` here, and reporting on zulip how it goes?

Reply:
Should the same then apply earlier in lines 65-67?
```
/-- Porting Note: Added -- needed to make coe_id work-/
instance {X Y : SemiNormedGroup} : CoeFun (X ⟶ Y) fun _ => X → Y where
  coe (f : X ⟶ Y) := NormedAddGroupHom.toFun f`
  ```

Reply:
Yes please!

Reply:
Should the instances be named `funLike` or `FunLike_instance`? I also asked [here](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/FunLike.20issues/near/365274729).

Reply:
I tried adding `FunLike` and `ZeroHomClass` instances for both `SemiNormedGroupCat` and `SemiNormedGroupCat₁`. Ready for review.